### PR TITLE
Implement module for CloudWatch logs

### PR DESCRIPTION
**Pull request by SweetOps**
## What

* Implement module for `CloudWatch` logs

## Why

* We need that for collection logs from instances

### CODE REVIEW

**Comment by SweetOps** [comment]
screenshot 2017-11-02 17 25 30

screenshot 2017-11-02 17 29 13

File: iam_user.tf [hunk 1]
```diff
@@ -0,0 +1,43 @@
+resource "aws_iam_user" "default" {
```
Comment (osterman) [inline iam_user.tf]: Can we use this module instead: https://github.com/cloudposse/terraform-aws-iam-system-user

Comment (SweetOps) [reply iam_user.tf]: @osterman,  in my case user will be create if `username` is defined,  with module `terraform-aws-iam-system-user` can't get same functional. The one way is add `count` to module `terraform-aws-iam-system-user`.

Comment (SweetOps) [reply iam_user.tf]: @osterman , I made PR in https://github.com/cloudposse/terraform-aws-iam-system-user/pull/2#issuecomment-342165056

File: iam_user.tf [hunk 2]
```diff
@@ -0,0 +1,33 @@
+data "aws_iam_policy_document" "user" {
+  count = "${var.create_user == "true" ? 1 : 0}"
```
Comment (osterman) [inline iam_user.tf]: let's rename to `user_enabled`

File: variables.tf [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  default     = ""
+}
+
+variable "retention_in_days" {
+  description = "Number of days you want to retain log events in the log group"
+  default     = "30"
+}
+
+variable "stream_names" {
+  default     = ["pfsense", "ubuntu"]
```
Comment (osterman) [inline variables.tf]: `pfsense` is a very specific use case. It doesn't make sense as a global default.

Comment (osterman) [reply variables.tf]: let's just create a default stream with the name of `default`

**Comment by osterman** [comment]
I activated TravisCI. Please bump.

File: iam.tf [hunk 4]
```diff
@@ -0,0 +1,40 @@
+data "aws_iam_policy_document" "log_assume" {
+  statement {
+    actions = ["sts:AssumeRole"]
+
+    principals {
+      type        = "Service"
+      identifiers = ["logs.${length(var.region) > 0 ? var.region: data.aws_region.default.name}.amazonaws.com"]
```
Comment (comeanother) [inline iam.tf]: Document `var.region` in README.

File: iam.tf [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      "logs:CreateLogGroup",
+      "logs:CreateLogStream",
+      "logs:PutLogEvents",
+      "logs:DescribeLogGroups",
+      "logs:DescribeLogStreams",
+      "logs:CreateLogStream",
+      "logs:DeleteLogStream",
+    ]
+
+    resources = [
```
Comment (comeanother) [inline iam.tf]: Use `compact` with `concat`.

File: iam.tf [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+      "${aws_cloudwatch_log_group.default.arn}",
+      "${join(",", aws_cloudwatch_log_stream.default.*.arn)}",
```
Comment (comeanother) [inline iam.tf]: Is `join(","` deliberate?

File: main.tf [hunk 7]
```diff
@@ -0,0 +1,35 @@
+data "aws_region" "default" {
+  current = true
```
Comment (comeanother) [inline main.tf]: `"true"`

File: iam.tf (hunk 6 above)
Comment (SweetOps) [reply iam.tf]: yes, for prevent mistakes in `json`

File: iam.tf (hunk 5 above)
Comment (SweetOps) [reply iam.tf]: > Use compact with concat.

It can break the module work - with `json` need be very careful

**Comment by comeanother** [comment]
@SweetOps @osterman 
Should we rename module to `terraform-aws-cloudwatch-log-stream` for consistence?

File: iam.tf (hunk 5 above)
Comment (osterman) [reply iam.tf]: Explain why it can break json

File: iam.tf (hunk 6 above)
Comment (osterman) [reply iam.tf]: explain what mistakes it prevents. it's not clear.

Comment (SweetOps) [reply iam.tf]: element in `resources` should be separated by comma, I'm not sure that comma will be without `join()`

File: iam.tf (hunk 5 above)
Comment (SweetOps) [reply iam.tf]: I'm sure which delimiter between elements I'll get after the `concat()`

**Comment by SweetOps** [comment]
@comeanother , I'm prefer the current name of repo :)

File: variables.tf [hunk 8]
```diff
@@ -0,0 +1,65 @@
+variable "name" {
+  default     = ""
+  description = "Name  (e.g. `bastion` or `db`)"
+}
+
+variable "namespace" {
+  description = "Namespace (e.g. `cp` or `cloudposse`)"
+  default     = "cp"
```
Comment (comeanother) [inline variables.tf]: Drop default value.

File: variables.tf [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+  type        = "string"
+}
+
+variable "stage" {
+  default     = "dev"
```
Comment (comeanother) [inline variables.tf]: Drop default value.

File: iam_user.tf [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+      "logs:DescribeLogStreams",
+      "logs:CreateLogStream",
+      "logs:DeleteLogStream",
+    ]
+
+    resources = ["*"]
+  }
+}
+
+module "user" {
```
Comment (comeanother) [inline iam_user.tf]: Do we need this module?
It is not in use.

File: iam.tf (hunk 5 above)
Comment (comeanother) [reply iam.tf]: It seems, there is no need to use `join` with delimiter here.
https://www.terraform.io/docs/providers/aws/d/iam_policy_document.html

File: iam_user.tf (hunk 10 above)
Comment (SweetOps) [reply iam_user.tf]: >It is not in use.
 
Why do you decided that?

**Comment by SweetOps** [comment]
> It seems, there is no need to use join with delimiter here.
> https://www.terraform.io/docs/providers/aws/d/iam_policy_document.html

@comeanother , I see in example resources separated by comma

File: outputs.tf [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+output "secret_access_key" {
+  description = "Secret access key"
+  value       = "${module.user.secret_access_key}"
+}
+
+output "log_group_name" {
+  value = "${aws_cloudwatch_log_group.default.name}"
+}
+
+output "log_stream_name" {
```
Comment (comeanother) [inline outputs.tf]: `log_stream_names`.

File: outputs.tf [hunk 12]
```diff
@@ -0,0 +1,42 @@
+output "log_group_arn" {
+  value       = "${aws_cloudwatch_log_group.default.arn}"
+  description = "ARN of the log group"
+}
+
+output "stream_arn" {
```
Comment (comeanother) [inline outputs.tf]: `stream_arns`.

File: main.tf [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+  current = "true"
+}
+
+module "log_group_label" {
+  source     = "git::https://github.com/cloudposse/terraform-null-label.git?ref=tags/0.3.1"
+  namespace  = "${var.namespace}"
+  name       = "${var.name}"
+  stage      = "${var.stage}"
+  delimiter  = "${var.delimiter}"
+  attributes = "${compact(concat(var.attributes, list("log")))}"
```
Comment (comeanother) [inline main.tf]: Should we add `group` attribute?

File: iam_user.tf (hunk 10 above)
Comment (comeanother) [reply iam_user.tf]: I assume, purpose of this user is to be able to connect to the corresponding streams.
But this user would has access to all streams. https://github.com/cloudposse/terraform-aws-cloudwatch-logs/pull/1/files#diff-a83a8bb953f15707ac8beb549b114ab5R22

**Pull request merged**